python/pycrdt/_awareness.py

```diff
@@ -0,0 +1,182 @@
+from __future__ import annotations
+
+import json
+import time
+from typing import Any, Callable
+from uuid import uuid4
+
+from ._doc import Doc
+from ._sync import Decoder, YMessageType, read_message, write_var_uint
+
+DEFAULT_USER = {"username": str(uuid4()), "name": "Jupyter server"}
+
+
+class Awareness:
+    client_id: int
+    meta: dict[int, dict[str, Any]]
+    _states: dict[int, dict[str, Any]]
+    _subscriptions: list[Callable[[dict[str, Any]], None]]
+    _user: dict[str, str] | None
+
+    def __init__(
+        self,
+        ydoc: Doc,
+        on_change: Callable[[bytes], None] | None = None,
+        user: dict[str, str] | None = None,
+    ):
+        self.client_id = ydoc.client_id
+        self.meta = {}
+        self._states = {}
+        self.on_change = on_change
+
+        if user is not None:
+            self.user = user
+        else:
+            self._user = DEFAULT_USER
+            self._states[self.client_id] = {"user": DEFAULT_USER}
+
+        self._subscriptions = []
+
+    @property
+    def states(self) -> dict[int, dict[str, Any]]:
+        return self._states
+
+    @property
+    def user(self) -> dict[str, str] | None:
+        return self._user
+
+    @user.setter
+    def user(self, user: dict[str, str]):
+        self._user = user
+        self.set_local_state_field("user", self._user)
+
+    def get_changes(self, message: bytes) -> dict[str, Any]:
```

Comment:
This method doesn't exist in [awareness.js](https://github.com/yjs/y-protocols/blob/255b7e6c96d9c6dd48b5519b3f83ab65afdb4bba/awareness.js). Do you think we could model the Python `Awareness` like the JavaScript `Awareness`?

Reply:
I think this method is equivalent of [`applyAwarenessUpdate`](https://github.com/yjs/y-protocols/blob/40dbe4eebb1e53a7e86932ef3232f9abd5037569/awareness.js#L241). We can rename it for consistency.

Reply:
But then it will not be backward compatible, since this method is already used by [jupyter_collaboration](https://github.com/jupyterlab/jupyter-collaboration/blob/b68d3cd756f3cc27bad447d5a6dbe1d5207c87dd/projects/jupyter-server-ydoc/jupyter_server_ydoc/handlers.py#L292), at least.